### ISSUE TITLE
Labelled do parsing error

### ISSUE THREAD

**Issue by RaoulHC** [camfort/fortran-src#155]
The following example fails to parse
```fortran
      subroutine foo
      integer ii
      do 10 ii=1,5
10    end do
      end
```

Looks like this bug was introduced in 9cce83fcca605392f9ea0bfeb11a2553044bf9f2, so I guess the `groupDo` and `groupLabelledDo` transformations were not commutative though I'm not entirely sure why.

Currently looking into it, but thought I'd raise an issue in case anyone had some good pointers.

**Comment by RaoulHC** [comment]
Forgot to put the output:
```
ProgramFilefortran-src: Mis-matched grouping statements at these position(s): do-loop-bug.f (5:7)-(5:12)
CallStack (from HasCallStack):
  error, called at src/Language/Fortran/Transformation/Grouping.hs:27:14 in fortran-src-0.4.3-CWQaufo8MNcESqc7843vtV:Language.Fortran.Transformation.Grouping
```

I'll try and fix this in #154, but if it's decided to make this optional this issue will still need to be resolved.

**Comment by raehik** [comment]
I came across the same issue when discussing #154 with Dom -- I think it's due to how grouping functions are implemented. They all call `genericGroup` with a block transform function (map), and a post-transform check (filter). The unlabeled do transformation uses:

```haskell
isDo :: Statement a -> Bool
isDo s = case s of
  StDo _ _ _ Nothing _      -> True
  StDoWhile _ _ _ Nothing _ -> True
  StEnddo{}                 -> True
  _                         -> False
```

Changing `end do` to `continue` allows parsing and transformation to complete with no errors. I don't think this is exactly what's happening because your example is a labeled do. But that error message comes from the checking function.

I'm writing more edge cases to compare between #154 and master. We should try disabling some of the transformations, see what happens. I *think* they should be idempotent, so assuming the PR implements them properly they *should* be a simple no-op when run again, but weird code always complicates things.

**Comment by raehik** [comment]
Didn't realise my mistake until I was going through related code! You're right, due to the filter check, they have to be run in a certain order. Fixed/reverted in #162 with a note on usage.
